core/engine/src/value/conversions/nullable.rs

```diff
@@ -0,0 +1,314 @@
+    fn try_into_js(&self, context: &mut Context) -> JsResult<JsValue> {
+        match self {
+            Nullable::Null => Ok(JsValue::null()),
+            Nullable::NonNull(t) => t.try_into_js(context),
+        }
+    }
+}
+
+impl<T> From<Nullable<T>> for Option<T> {
+    #[inline]
+    fn from(value: Nullable<T>) -> Self {
+        match value {
+            Nullable::Null => None,
+            Nullable::NonNull(t) => Some(t),
+        }
+    }
+}
+
+impl<T> From<Option<T>> for Nullable<T> {
+    #[inline]
+    fn from(value: Option<T>) -> Self {
+        match value {
+            None => Nullable::Null,
+            Some(t) => Nullable::NonNull(t),
+        }
+    }
+}
+
+/// An iterator over a [`Nullable`] value reference.
+#[derive(Debug)]
+pub struct Iter<'a, A>(Option<&'a A>);
+
+impl<'a, T: 'a> Iterator for Iter<'a, T> {
+    type Item = &'a T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.0.take()
+    }
+}
+
+impl<'a, T: 'a> DoubleEndedIterator for Iter<'a, T> {
+    fn next_back(&mut self) -> Option<Self::Item> {
+        self.0.take()
+    }
+}
+
+impl<A> ExactSizeIterator for Iter<'_, A> {
+    #[inline]
+    fn len(&self) -> usize {
+        usize::from(self.0.is_some())
+    }
+}
+
+/// An owning iterator over a [`Nullable`] value.
+#[derive(Debug)]
+pub struct IntoIter<A>(Option<A>);
+
+impl<A> Iterator for IntoIter<A> {
+    type Item = A;
+    fn next(&mut self) -> Option<Self::Item> {
+        self.0.take()
+    }
+}
+
+impl<A> DoubleEndedIterator for IntoIter<A> {
+    fn next_back(&mut self) -> Option<A> {
+        self.0.take()
+    }
+}
+
+impl<A> ExactSizeIterator for IntoIter<A> {
+    #[inline]
+    fn len(&self) -> usize {
+        usize::from(self.0.is_some())
+    }
+}
+
+impl<T> Nullable<Option<T>> {
+    /// Converts from `Nullable<Option<T>>` to `Option<T>`.
+    #[inline]
+    pub fn flatten(self) -> Option<T> {
+        match self {
+            Nullable::Null | Nullable::NonNull(None) => None,
+            Nullable::NonNull(Some(t)) => Some(t),
+        }
+    }
+}
+
+#[test]
+fn not_null() {
+    let context = &mut Context::default();
+    let v: Nullable<i32> = JsValue::new(42)
+        .try_js_into(context)
+        .expect("Failed to convert value from js");
+
+    assert!(!v.is_null());
+    assert!(v.is_not_null());
+    assert_eq!(v, Nullable::NonNull(42));
+
+    assert_eq!(v.try_into_js(context).unwrap(), JsValue::new(42));
+}
+
+#[test]
+fn null() {
+    let context = &mut Context::default();
+    let v: Nullable<i32> = JsValue::null()
+        .try_js_into(context)
+        .expect("Failed to convert value from js");
+
+    assert!(v.is_null());
+    assert!(!v.is_not_null());
+    assert_eq!(v, Nullable::Null);
+
+    assert_eq!(v.try_into_js(context).unwrap(), JsValue::null());
+}
+
+#[test]
+fn invalid() {
+    let context = &mut Context::default();
+    let v: JsResult<Nullable<i32>> = JsValue::undefined().try_js_into(context);
+
+    assert!(v.is_err());
+}
```

Comment:
consider putting this in a `#[cfg(test)]` `tests` module in-file or as a separate `tests.rs` module.